crates/biome_module_graph/src/js_semantic_model.rs

```diff
@@ -0,0 +1,309 @@
+use biome_js_syntax::{
+    AnyJsDeclaration, AnyJsImportClause, JsFunctionDeclaration, JsSyntaxNode,
+    JsVariableDeclaration, JsVariableKind, TsTypeAliasDeclaration,
+};
+use biome_js_type_info::Type;
+use biome_rowan::{AstNode, Text};
+
+use crate::jsdoc_comment::JsdocComment;
+
+/// Simplified version of the JS Semantic Model, that provides just enough
+/// information to keep track of the types of all symbols in a module's global
+/// scope.
+///
+/// The model is discarded as soon as the module's inferrence has completed.
+#[derive(Clone, Debug)]
+pub(crate) struct JsSemanticModel {
+    pub declarations: Box<[JsDeclaration]>,
+}
+
+impl JsSemanticModel {
+    /// Finds the value declaration with the given `name`.
+    pub fn get_value(&self, name: &str) -> Option<&JsDeclaration> {
+        self.declarations
+            .iter()
+            .find(|decl| decl.name == name && decl.kind.might_be_value())
+    }
+
+    /// Finds the type declaration with the given `name`.
+    pub fn get_type(&self, name: &str) -> Option<&JsDeclaration> {
+        self.declarations
+            .iter()
+            .find(|decl| decl.name == name && decl.kind.might_be_type())
+    }
+}
+
+/// A single declaration of a variable or type.
+#[derive(Clone, Debug)]
+pub struct JsDeclaration {
+    /// The name of the thing being declared.
+    pub name: Text,
+
+    /// The kind of thing being declared.
+    pub kind: JsDeclarationKind,
+
+    /// Optional JSDoc comment associated with the declaration.
+    pub jsdoc_comment: Option<JsdocComment>,
+}
+
+#[derive(Clone, Debug)]
+pub enum JsDeclarationKind {
+    /// A `class` declaration.
+    ///
+    /// The type defined by the class.
+    #[expect(unused)] // TODO
+    Class(Type),
+
+    /// A `function` or `var` declaration.
+    ///
+    /// The type (either declared or inferred) of the value.
+    HoistedValue(Type),
+
+    /// An `import` declaration.
+    Import(Text),
+
+    /// An `import type` declaration.
+    ImportType(Text),
+
+    /// An interface declaration.
+    #[expect(unused)] // TODO
+    Interface,
+
+    /// A module declaration.
+    #[expect(unused)] // TODO
+    Module,
+
+    /// A namespace declaration.
+    #[expect(unused)] // TODO
+    Namespace,
+
+    /// A type declaration.
+    Type(Type),
+
+    /// A `let` or `const` declaration.
+    ///
+    /// The type (either declared or inferred) of the value.
+    Value(Type),
+}
+
+impl JsDeclarationKind {
+    /// Returns `true` for any declaration that _may_ be a type.
+    fn might_be_type(&self) -> bool {
```

Comment:
I fail to understand the semantic difference between "might be a type" (`might_be_type`) and "is a type" (`is_a_type`), because a "might" could be translated into a `enum` where "maybe" is a valid variant.

Reply:
It's mainly due to the presence of the `Import` variant, where we don't know whether it's a type or a value being imported.

I'll clarify it with a comment.